### ISSUE TITLE
Resolved ts in stale read may cause TiKV OOM

### ISSUE THREAD

**Issue by overvenus** [tikv/tikv#14864]
## Bug Report

Stale read is enabled by default, and it maintains resolved ts(a btree map) for tracking every lock written by leaders.
If there is a large txn, the map may cause TiKV OOM.

To prevent OOM, we can use a memory quota to keep the total memory consumption under control. If a region is tracking too many locks and exceeds the quota, we can deregister its delegate in the stale read component, free its memory, and then register it again after a certain backoff period.

### What version of TiKV are you using?

\> 5.1.0

### Steps to reproduce

Run a large txn that is much greater than TiKV memory size.

### What did you expect?

No OOM.

### What did happened?

OOM.

**Comment by tonyxuqqi** [comment]
@overvenus Do we have metrics for the memory consumption (direct or indirect) of the resolved-ts map? It can be the first step to solve this issue

**Comment by overvenus** [comment]
Yes, we have such metric, it's TiKV-Details -> Resolved-TS -> Lock heap size

![image](https://github.com/tikv/tikv/assets/2150711/3bdc92cc-39da-43a0-bc66-e9ec24239b65)

**Comment by tonyxuqqi** [comment]
The overall solution looks good to me. One detail is that maybe we can use the free memory ratio as the config to evict some large region's entry, other than using a quota. Since it's hard to ensure all components' quota added up to be 100% of total memory.
